src/trackers.js

```diff
@@ -1,227 +1,255 @@
-            // Ex: x.y.z.analytics.com would be checked 4 times pulling off a subdomain each time.
-            url.shift()
-            return this.checkTrackersWithParentCompany(url, currLocation)
+        // look up surrogate
+        if (matchedRule && matchedRule.surrogate) {
+            tracker.redirectUrl = this.surrogateList[matchedRule.surrogate]
         }
+        return matchedRule
     }
 
-    /* Check to see if this tracker is related to the current page through their parent companies
-     * Only block request to 3rd parties
-     */
-    isRelatedEntity (parentCompany, currLocation) {
-        let parentEntity = this.entityList[parentCompany]
-        let host = utils.extractHostFromURL(currLocation)
-
-        if (parentEntity && parentEntity.properties) {
-        // join parent entities to use as regex and store in parentEntity so we don't have to do this again
-            if (!parentEntity.regexProperties) {
-                parentEntity.regexProperties = parentEntity.properties.join('|')
-            }
-
-            if (host.match(parentEntity.regexProperties)) {
-                return true
-            }
+    requestMatchesRule (requestData, ruleObj) {
+        if (requestData.urlToCheck.match(ruleObj.rule)) {
+            return this.matchRuleOptions(ruleObj, requestData)
+        } else {
+            return false
         }
-
-        return false
     }
 
-    /* Compare two urls to determine if they came from the same hostname
-     * pull off any subdomains before comparison.
-     * Return parent company name from entityMap if one is found or unknown
-     * if domains match but we don't have this site in our entityMap.
-     */
-    getCommonParentEntity (currLocation, urlToCheck) {
-        if (!entityMap) return
-        let currLocationDomain = utils.getDomain(currLocation)
-        let urlToCheckDomain = utils.getDomain(urlToCheck)
-        let parentEntity = entityMap[urlToCheckDomain]
-        if (currLocationDomain === urlToCheckDomain ||
-                this.isRelatedEntity(parentEntity, currLocation)) {
-            return parentEntity || currLocationDomain
+    /* Check the matched rule exceptions against the request data
+     * Both domains and types need to match (if they exist) in order
+     * for the exception to match.
+    */
+    matchesException (tracker, requestData) {
+        if (tracker.matchedRule && tracker.matchedRule.exceptions) {
+            // all exception types need to match for this to return true
+            return !this.exceptionTypes.some(exceptionType => {
+                const requestParamToCheck = exceptionType === 'domains' ? requestData.siteDomain : requestData.request.type
+                return !(tracker.matchedRule.exceptions[exceptionType] &&
+                          tracker.matchedRule.exceptions[exceptionType].length &&
+                          tracker.matchedRule.exceptions[exceptionType].includes(requestParamToCheck))
+            })
```

Comment:
this is pretty much the same as `matchRuleOptions`, can we reuse `matchRuleOptions`?

Reply:
Added a more general function `matchesRuleDefinition` that handles both options and exceptions.